components/backup/src/disk_snap.rs

```diff
@@ -0,0 +1,340 @@
+    }
+}
+
+enum StreamHandleEvent {
+    Req(PReq),
+    WaitApplyDone(Region, Result<()>),
+    ConnectionGone(Option<grpcio::Error>),
+}
+
+impl<SR: SnapshotBrHandle + 'static> StreamHandleLoop<SR> {
+    pub fn new(env: Env<SR>) -> Self {
+        env.active_stream.fetch_add(1, Ordering::SeqCst);
+        Self {
+            env,
+            pending_regions: vec![],
+        }
+    }
+
+    fn async_wait_apply(&mut self, region: &Region) -> BoxFuture<'static, (Region, Result<()>)> {
+        if let Err(err) = self.env.check_rejected() {
+            return Box::pin(future::ready((region.clone(), Err(err))));
+        }
+
+        let (tx, rx) = oneshot::channel();
+        let syncer = SnapshotBrWaitApplySyncer::new(region.id, tx);
+        let handle = self.env.handle.clone();
+        let region = region.clone();
+        let epoch = region.get_region_epoch().clone();
+        let id = region.get_id();
+        let send_res = handle
+            .send_wait_apply(id, SnapshotBrWaitApplyRequest::strict(syncer, epoch))
+            .map_err(Error::RaftStore);
+        Box::pin(
+            async move {
+                send_res?;
+                rx.await.map_err(|_| Error::WaitApplyAborted).map(|_| ())
+            }
+            .map(move |res| (region, res)),
+        )
+    }
+
+    async fn next_event(
+        &mut self,
+        input: &mut (impl Stream<Item = grpcio::Result<PReq>> + Unpin),
+    ) -> StreamHandleEvent {
+        let wait_applies = future::poll_fn(|cx| {
+            let selected =
+                self.pending_regions
+                    .iter_mut()
+                    .enumerate()
+                    .find_map(|(i, fut)| match fut.poll_unpin(cx) {
+                        Poll::Ready(r) => Some((i, r)),
+                        Poll::Pending => None,
+                    });
+            match selected {
+                Some((i, region)) => {
+                    self.pending_regions.swap_remove(i);
+                    region.into()
+                }
+                None => Poll::Pending,
+            }
+        });
+
+        tokio::select! {
+            wres = wait_applies => {
+                StreamHandleEvent::WaitApplyDone(wres.0, wres.1)
+            }
+            req = input.next() => {
+                match req {
+                    Some(Ok(req)) => StreamHandleEvent::Req(req),
+                    Some(Err(err)) => StreamHandleEvent::ConnectionGone(Some(err)),
+                    None => StreamHandleEvent::ConnectionGone(None)
+                }
+            }
+        }
+    }
+
+    pub async fn run(
+        mut self,
+        mut input: impl Stream<Item = grpcio::Result<PReq>> + Unpin,
+        mut sink: ResultSink,
+    ) -> grpcio::Result<()> {
+        loop {
+            match self.next_event(&mut input).await {
+                StreamHandleEvent::Req(req) => match req.get_ty() {
+                    PReqT::UpdateLease => {
```

Comment:
Could you explain somewhere what's problem that lease is to solve

Reply:
I guess perhaps we can write this to kvproto...?